rise_set/visibility.py

```diff
@@ -233,6 +233,43 @@ def get_zenith_distance_intervals(self, target, target_intervals, chunksize=date
         """
         return self._get_chunked_intervals(target, target_intervals, self._add_zenith_interval, self.zenith_blind_spot, chunksize)
 
+    def get_moon_phase_intervals(self, target_intervals, max_moon_phase=1.0, chunksize=datetime.timedelta(minutes=30)):
+        """ Returns the intervals in which the target is visible and the moon phase does not exceed max_moon_phase
+
+        Args:
+            target_intervals (list): A list of datetime tuples that represent the above horizon intervals for the target.
+                                     Returned by get_target_intervals()
+            max_moon_phase (float): max allowablable fractional moon phase from 0 to 1 (full moon).
+            chunksize (timedelta): The time delta over which to calculate if the target intervals are out of range of
+                                   the zenith.
+        Returns:
+            list: A list of tuples of start/end datetime pairs that make up the intervals over which this target is visible and the moon phase is less than max_moon_phase
+        """
+        # If target intervals are empty, no point in calculating moon phases
+        if not target_intervals:
+            return target_intervals
+        bad_intervals = []
+        moon_up_intervals = self.get_target_intervals(target="moon", up=True)
+        target_moon_up_intervals = intersect_intervals(target_intervals, moon_up_intervals)
+        # Here we collect the bad intervals where the moon is up and the moon phase is greater than the max constraint
+        for start, end in target_moon_up_intervals:
+            chunkstart = start
+            chunkend = min(chunkstart + chunksize, end)
+            while chunkstart != chunkend and chunkend <= end:
+                moon_phase = calculate_moon_phase(chunkstart, self.site['latitude'].in_radians(), self.site['longitude'].in_radians())
+                if moon_phase > max_moon_phase:
+                    bad_intervals.append((chunkstart, chunkend))
+
+                # increment the chunkstart/end up
+                chunkstart = chunkend
+                chunkend = min(chunkstart + chunksize, end)
+
+        bad_intervals = coalesce_adjacent_intervals(bad_intervals)
+        # Bad intervals are inversed to get good intervals and then intersected with target intervals to get target intervals where moon_phase constraint is met
+        good_intervals = inverse_intervals(bad_intervals, target_intervals[0][0], target_intervals[-1][1])
+        good_intervals = intersect_intervals(good_intervals, target_intervals)
+        return good_intervals
```

Comment:
This function is a bit convoluted with calculating the bad intervals and then inversing to get good intervals, but I couldn't think of a better way to efficiently calculate the intervals over which moon is up AND moon phase > max.